src/main/java/org/jenkinsci/remoting/util/VersionNumber.java

```diff
@@ -0,0 +1,579 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.jenkinsci.remoting.util;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Locale;
+import java.util.Properties;
+import java.util.Stack;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Immutable representation of a version number based on the Mercury version numbering scheme.
+ *
+ * {@link VersionNumber}s are {@link Comparable}.
+ *
+ * <h2>Special tokens</h2>
+ * <p>
+ * We allow a component to be not just a number, but also "ea", "ea1", "ea2".
+ * "ea" is treated as "ea0", and eaN &lt; M for any M &lt; 0.
+ *
+ * <p>
+ * '*' is also allowed as a component, and '*' &lt; M for any M &lt; 0.
+ *
+ * <p>
+ * 'SNAPSHOT' is also allowed as a component, and "N.SNAPSHOT" is interpreted as "N-1.*"
+ *
+ * <pre>
+ * 2.0.* &lt; 2.0.1 &lt; 2.0.1-SNAPSHOT &lt; 2.0.0.99 &lt; 2.0.0 &lt; 2.0.ea &lt; 2.0
+ * </pre>
+ *
+ * This class is re-implemented in 1.415. The class was originally introduced in 1.139
+ *
+ * @since 1.139
+ * @author Stephen Connolly (stephenc@apache.org)
+ * @author Kenney Westerhof (kenney@apache.org)
+ * @author Hervé Boutemy (hboutemy@apache.org)
+ */
+public class VersionNumber implements Comparable<VersionNumber> {
```

Comment:
This class should be restricted from external use, since we really don't want to provide it as part of the API for Remoting. In other words, add the annotation, `@Restricted(NoExternalUse.class)` to the class.

Also, could you leave a comment about where this was copied from?